File: Sources/ClaudeNotch/StatuslineHook.swift
import Foundation

/// 与 Claude Code 的 `statusLine` 钩子对接，取「官方/合规」的订阅额度。
///
/// 背景：消费者订阅（Pro/Max）的 5 小时 / 周额度**没有公开 API**。Claude Code 自己是从
/// 每次 `/v1/messages` 响应头 `anthropic-ratelimit-unified-*` 拿到的，并会把它整理成
/// `rate_limits.{five_hour,seven_day}`（`used_percentage` 0–100、`resets_at` Unix 秒）
/// 通过 **stdin 喂给用户配置的 statusLine 命令**——这是官方文档化的第三方钩子契约。
/// 复用 OAuth 令牌打 `/api/oauth/usage` 是 Anthropic 明令禁止（2026-02 起）且会被限流的做法；
/// 抓网页同样属自动化访问。相比之下，让 Claude Code **主动把数据交给我们**，是合规风险最低的实时来源。
///
/// 机制：
/// 1. 启用时，把本 app 自身（`<bin> --statusline`）写进 `~/.claude/settings.json` 的 `statusLine`，
///    并备份、链接用户原有的 statusline（透传，不抢占）。
/// 2. Claude Code 渲染状态栏时以 `--statusline` 调起本 app；`runHelper()` 读 stdin、
///    把 `rate_limits` 落盘到 `ratelimits.json`，再把 stdin 转发给原命令并输出其结果。
/// 3. `StatuslineProvider` 读 `ratelimits.json` 归一成 `ScrapeResult`。
///
/// 全程不读取、不复用任何令牌；只写自己的支持目录与（用户显式开启时）`~/.claude/settings.json`。
enum StatuslineHook {

    // MARK: - 共享路径

    static var supportDir: URL {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
            ?? FileManager.default.homeDirectoryForCurrentUser.appendingPathComponent("Library/Application Support")
        return base.appendingPathComponent("ClaudeNotch", isDirectory: true)
    }
    /// 钩子落盘的额度数据，`StatuslineProvider` 读它。
    static var ratelimitsFile: URL { supportDir.appendingPathComponent("ratelimits.json") }
    /// 接管前用户原有的 statusLine **整个对象**（用于透传 + 卸载时原样还原，保留 padding 等所有字段）。
    static var innerStatusLineFile: URL { supportDir.appendingPathComponent("inner-statusline.json") }
    static var claudeSettings: URL {
        FileManager.default.homeDirectoryForCurrentUser.appendingPathComponent(".claude/settings.json")
    }

    /// 判断某条 statusLine 命令是否是我们自己（避免把自己存成「原命令」造成自我递归）。
    private static func isOurs(_ command: String) -> Bool {
        command.contains("--statusline") && command.contains("ClaudeNotch")
    }

    // MARK: - 作为 statusLine 命令运行（main.swift 检测到 --statusline 时调用）

    /// 读 stdin、落盘额度、透传原 statusline。必须快进快出，不启动任何 GUI。
    static func runHelper() {
        let input = FileHandle.standardInput.readDataToEndOfFile()
        let root = (try? JSONSerialization.jsonObject(with: input)) as? [String: Any]
        let rateLimits = root?["rate_limits"] as? [String: Any]

        if let rl = rateLimits { persist(rateLimits: rl) }

        // 透传：有原命令就转发同一份 stdin 并原样输出其状态栏；否则打印一行简洁默认。
        if let inner = innerCommand() {
            forward(input: input, command: inner, rateLimitsFallback: rateLimits)
        } else {
            FileHandle.standardOutput.write(Data(defaultLine(root: root, rateLimits: rateLimits).utf8))
        }
    }

    private static func persist(rateLimits: [String: Any]) {
        let payload: [String: Any] = [
            "capturedAt": Date().timeIntervalSince1970,
            "rate_limits": rateLimits,
        ]
        try? FileManager.default.createDirectory(at: supportDir, withIntermediateDirectories: true)
        if let data = try? JSONSerialization.data(withJSONObject: payload) {
            try? data.write(to: ratelimitsFile, options: .atomic)
        }
    }

    /// 接管前的原 statusLine 对象（整体，用于卸载还原）。
    private static func innerStatusLine() -> [String: Any]? {
        guard let data = try? Data(contentsOf: innerStatusLineFile),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return nil }
        return obj
    }

    /// 从原 statusLine 对象里取出可执行命令（用于透传）。
    private static func innerCommand() -> String? {
        guard let cmd = innerStatusLine()?["command"] as? String else { return nil }
        let t = cmd.trimmingCharacters(in: .whitespacesAndNewlines)
        return t.isEmpty ? nil : t
    }

    /// 把同一份 stdin 转发给原 statusline 命令，原样输出其 stdout。
    private static func forward(input: Data, command: String, rateLimitsFallback: [String: Any]?) {
        let p = Process()
        p.executableURL = URL(fileURLWithPath: "/bin/sh")
        p.arguments = ["-c", command]
        let inPipe = Pipe(), outPipe = Pipe()
        p.standardInput = inPipe
        p.standardOutput = outPipe
        p.standardError = FileHandle.standardError
        do {
            try p.run()
            inPipe.fileHandleForWriting.write(input)
            try? inPipe.fileHandleForWriting.close()
            let out = outPipe.fileHandleForReading.readDataToEndOfFile()
            p.waitUntilExit()
            FileHandle.standardOutput.write(out)
        } catch {
            FileHandle.standardOutput.write(Data(defaultLine(root: nil, rateLimits: rateLimitsFallback).utf8))
        }
    }

    /// 没有原 statusline 时的默认行：模型名 · 5h% · 7d%（避免 CLI 状态栏变空）。
    private static func defaultLine(root: [String: Any]?, rateLimits: [String: Any]?) -> String {
        var parts: [String] = []
        if let model = root?["model"] as? [String: Any], let name = model["display_name"] as? String {
            parts.append(name)
        }
        func pct(_ key: String, _ label: String) {
            if let w = rateLimits?[key] as? [String: Any], let p = (w["used_percentage"] as? NSNumber)?.intValue {
                parts.append("\(label) \(p)%")
            }
        }
        pct("five_hour", "5h")
        pct("seven_day", "7d")
        return parts.joined(separator: " · ")
    }

    // MARK: - 安装 / 卸载

    /// 幂等确保已注册（启动时调用）：未安装、或已安装但指向的二进制不是当前 app（app 被移动/换构建）时，(重新)安装。
    static func ensureInstalled() {
        guard let bin = Bundle.main.executablePath else { return }
        if isInstalled, currentCommand()?.contains(bin) == true { return }
        install()
    }

    private static func currentCommand() -> String? {
        guard let data = try? Data(contentsOf: claudeSettings),
              let s = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let sl = s["statusLine"] as? [String: Any] else { return nil }
        return sl["command"] as? String
    }

    /// 把本 app 注册为 Claude Code 的 statusLine 命令。会先整文件备份一次，并链接用户原有 statusline。
    static func install() {
        let fm = FileManager.default
        try? fm.createDirectory(at: supportDir, withIntermediateDirectories: true)
        guard let binPath = Bundle.main.executablePath else { return }

        var settings: [String: Any] = [:]
        if let data = try? Data(contentsOf: claudeSettings) {
            settings = (try? JSONSerialization.jsonObject(with: data) as? [String: Any]) ?? [:]
            // 整文件备份一次（仅首次，不覆盖已有备份）
            let backup = claudeSettings.appendingPathExtension("claudenotch-bak")
            if !fm.fileExists(atPath: backup.path) { try? data.write(to: backup) }
        }

        // 记下接管前的原 statusLine 整个对象（仅当它不是我们自己），供透传与卸载原样还原
        if let sl = settings["statusLine"] as? [String: Any], !isOurs(sl["command"] as? String ?? ""),
           let data = try? JSONSerialization.data(withJSONObject: sl) {
            try? data.write(to: innerStatusLineFile, options: .atomic)
        }

        settings["statusLine"] = [
            "type": "command",
            "command": "\"\(binPath)\" --statusline",
            "padding": 0,
        ]
        writeSettings(settings)
    }

    /// 撤销注册：把 statusLine 恢复成用户原有命令（若无则移除该键）。
    /// `purgeData=false`（退出前还原用）保留已抓到的 `ratelimits.json`，下次启动可秒显上次额度。
    static func uninstall(purgeData: Bool = true) {
        let fm = FileManager.default
        if let data = try? Data(contentsOf: claudeSettings),
           var settings = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
           let sl = settings["statusLine"] as? [String: Any],
           let cmd = sl["command"] as? String, isOurs(cmd) {       // 只有当前确实是我们才动它
            if let inner = innerStatusLine() {
                settings["statusLine"] = inner          // 整对象原样还原（保留 padding 等所有字段）
            } else {
                settings.removeValue(forKey: "statusLine")
            }
            writeSettings(settings)
        }
        try? fm.removeItem(at: innerStatusLineFile)   // 已还原进 settings.json，链接文件不再需要
        if purgeData { try? fm.removeItem(at: ratelimitsFile) }
    }

    /// 当前 `~/.claude/settings.json` 的 statusLine 是否已是我们。
    static var isInstalled: Bool {
        guard let data = try? Data(contentsOf: claudeSettings),
              let settings = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let sl = settings["statusLine"] as? [String: Any],
              let cmd = sl["command"] as? String else { return false }
        return isOurs(cmd)
    }

    private static func writeSettings(_ settings: [String: Any]) {
        let dir = claudeSettings.deletingLastPathComponent()
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        if let out = try? JSONSerialization.data(withJSONObject: settings, options: [.prettyPrinted, .sortedKeys]) {
            try? out.write(to: claudeSettings, options: .atomic)
        }
    }
}

File: windows/ClaudeNotch/Core/StatuslineHook.cs
using System.Diagnostics;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace ClaudeNotch.Core;

/// <summary>
/// 与 Claude Code 的 statusLine 钩子对接（额度的唯一来源，合规、不复用令牌）。
/// 把本 app（exe --statusline）注册进 ~/.claude/settings.json 的 statusLine；Claude Code 渲染状态栏时
/// 经 stdin 把 rate_limits 喂进来，钩子落盘 ratelimits.json，再透传原命令。
/// </summary>
public static class StatuslineHook
{
    const string OwnerMarker = "_claudenotch";

    static bool IsOurs(string command) =>
        command.Contains("ClaudeNotch.Statusline") || (command.Contains("--statusline") && command.Contains("ClaudeNotch"));

    /// <summary>
    /// 注册用的 statusLine 命令：优先专用助手 exe，缺失时退回主 exe --statusline。
    /// Claude Code 在 Windows 用 **PowerShell** 执行 statusLine：**带引号的裸字符串只会被回显、不执行**，
    /// 故用「裸正斜杠路径」(PowerShell 会当命令执行；正斜杠在 PS/cmd/bash 都不被转义)。
    /// 路径含空格时用调用运算符 `& '...'`（PowerShell 下对带空格路径有效）。
    /// </summary>
    static string HookCommand()
    {
        bool helperExists = File.Exists(Paths.StatuslineHelperExe);
        var exe = (helperExists ? Paths.StatuslineHelperExe : Paths.ExePath).Replace('\\', '/');
        var args = helperExists ? "" : " --statusline";
        return exe.Contains(' ') ? $"& '{exe}'{args}" : $"{exe}{args}";
    }

    static bool ObjectIsOurs(JsonObject sl)
    {
        if (sl.TryGetPropertyValue(OwnerMarker, out var marker) && marker is JsonValue v && v.TryGetValue<bool>(out var b) && b)
            return true;
        return IsOurs((sl["command"]?.GetValue<string>()) ?? "");
    }

    // ── 作为 statusLine 命令运行（Program.Main 检测到 --statusline 时调用） ──

    public static void RunHelper()
    {
        string input;
        using (var stdin = Console.OpenStandardInput())
        using (var reader = new StreamReader(stdin, Encoding.UTF8))
            input = reader.ReadToEnd();

        JsonObject? root = null;
        try { root = JsonNode.Parse(input) as JsonObject; } catch { }
        var rateLimits = root?["rate_limits"] as JsonObject;

        if (rateLimits is not null) Persist(rateLimits, root);

        var inner = InnerCommand();
        if (inner is not null) Forward(input, inner, rateLimits);
        else Console.Out.Write(DefaultLine(root, rateLimits));
    }

    static void Persist(JsonObject rateLimits, JsonObject? root)
    {
        var payload = new JsonObject
        {
            ["capturedAt"] = DateTimeOffset.UtcNow.ToUnixTimeSeconds(),
            ["rate_limits"] = rateLimits.DeepClone(),
        };
        if (root is not null)
            foreach (var key in new[] { "cost", "model", "version", "workspace", "output_style", "session_id" })
                if (root[key] is JsonNode n) payload[key] = n.DeepClone();
        try
        {
            Directory.CreateDirectory(Paths.SupportDir);
            File.WriteAllText(Paths.RatelimitsFile, payload.ToJsonString());
        }
        catch { }
    }

    static JsonObject? InnerStatusLine()
    {
        try
        {
            if (File.Exists(Paths.InnerStatusLineFile))
                return JsonNode.Parse(File.ReadAllText(Paths.InnerStatusLineFile)) as JsonObject;
        }
        catch { }
        return null;
    }

    static string? InnerCommand()
    {
        var cmd = InnerStatusLine()?["command"]?.GetValue<string>()?.Trim();
        return string.IsNullOrEmpty(cmd) ? null : cmd;
    }

    static void Forward(string input, string command, JsonObject? rateLimitsFallback)
    {
        try
        {
            var psi = new ProcessStartInfo("cmd.exe", "/c " + command)
            {
                RedirectStandardInput = true,
                RedirectStandardOutput = true,
                UseShellExecute = false,
                CreateNoWindow = true,
            };
            var p = Process.Start(psi)!;
            p.StandardInput.Write(input);
            p.StandardInput.Close();
            var outp = p.StandardOutput.ReadToEnd();
            p.WaitForExit();
            Console.Out.Write(outp);
        }
        catch
        {
            Console.Out.Write(DefaultLine(null, rateLimitsFallback));
        }
    }

    static string DefaultLine(JsonObject? root, JsonObject? rateLimits)
    {
        var parts = new List<string>();
        if (root?["model"] is JsonObject model && model["display_name"]?.GetValue<string>() is string name)
            parts.Add(name);
        void Pct(string key, string label)
        {
            if (rateLimits?[key] is JsonObject w && w["used_percentage"] is JsonValue pv && pv.TryGetValue<double>(out var p))
                parts.Add($"{label} {(int)Math.Round(p)}%");
        }
        Pct("five_hour", "5h");
        Pct("seven_day", "7d");
        return string.Join(" · ", parts);
    }

    // ── 安装 / 卸载 ──

    public static void EnsureInstalled()
    {
        // 已安装且命令已指向“当前应注册的可执行文件”才跳过；否则重装。
        // 按正斜杠归一化比较（注册命令现用正斜杠裸路径），避免反斜杠/正斜杠差异导致每次启动都重写。
        bool helperExists = File.Exists(Paths.StatuslineHelperExe);
        var target = (helperExists ? Paths.StatuslineHelperExe : Paths.ExePath).Replace('\\', '/');
        var current = CurrentCommand()?.Replace('\\', '/');
        if (IsInstalled && current is not null && current.Contains(target)) return;
        Install();
    }

    public static string? CurrentCommand()
    {
        try
        {
            if (!File.Exists(Paths.ClaudeSettings)) return null;
            var root = JsonNode.Parse(File.ReadAllText(Paths.ClaudeSettings)) as JsonObject;
            return (root?["statusLine"] as JsonObject)?["command"]?.GetValue<string>();
        }
        catch { return null; }
    }

    public static void Install()
    {
        try { Directory.CreateDirectory(Paths.SupportDir); } catch { }

        JsonObject settings = new();
        if (File.Exists(Paths.ClaudeSettings))
        {
            try
            {
                var text = File.ReadAllText(Paths.ClaudeSettings);
                settings = JsonNode.Parse(text) as JsonObject ?? new JsonObject();
                var backup = Paths.ClaudeSettings + ".claudenotch-bak";
                if (!File.Exists(backup)) File.WriteAllText(backup, text);
            }
            catch { settings = new JsonObject(); }
        }

        if (settings["statusLine"] is JsonObject sl && !ObjectIsOurs(sl))
        {
            try { File.WriteAllText(Paths.InnerStatusLineFile, sl.ToJsonString()); } catch { }
        }

        settings["statusLine"] = new JsonObject
        {
            ["type"] = "command",
            ["command"] = HookCommand(),
            ["padding"] = 0,
            [OwnerMarker] = true,
        };
        WriteSettings(settings);
    }

    public static void Uninstall(bool purgeData = true)
    {
        try
        {
            if (File.Exists(Paths.ClaudeSettings)
                && JsonNode.Parse(File.ReadAllText(Paths.ClaudeSettings)) is JsonObject settings
                && settings["statusLine"] is JsonObject sl && ObjectIsOurs(sl))
            {
                var inner = InnerStatusLine();
                if (inner is not null) settings["statusLine"] = inner.DeepClone();
                else settings.Remove("statusLine");
                WriteSettings(settings);
                if (File.Exists(Paths.InnerStatusLineFile)) File.Delete(Paths.InnerStatusLineFile);
            }
        }
        catch { }
        if (purgeData) { try { if (File.Exists(Paths.RatelimitsFile)) File.Delete(Paths.RatelimitsFile); } catch { } }
    }

    public static bool IsInstalled
    {
        get
        {
            try
            {
                if (!File.Exists(Paths.ClaudeSettings)) return false;
                var settings = JsonNode.Parse(File.ReadAllText(Paths.ClaudeSettings)) as JsonObject;
                return settings?["statusLine"] is JsonObject sl && ObjectIsOurs(sl);
            }
            catch { return false; }
        }
    }

    static void WriteSettings(JsonObject settings)
    {
        try
        {
            Directory.CreateDirectory(Path.GetDirectoryName(Paths.ClaudeSettings)!);
            File.WriteAllText(Paths.ClaudeSettings, settings.ToJsonString(new JsonSerializerOptions { WriteIndented = true }));
        }
        catch { }
    }

    // ── 诊断（设置页用） ──

    public record Diagnostics(bool Installed, string? Command, string? WrappedInner,
        bool RatelimitsExists, DateTime? CapturedAt, string SettingsPath, string SupportDirPath)
    {
        public string CopyText()
        {
            var lines = new List<string>
            {
                L.Tr("ClaudeNotch 集成诊断", "ClaudeNotch integration diagnostics"),
                L.Tr("已接入: ", "Installed: ") + (Installed ? L.Tr("是", "Yes") : L.Tr("否", "No")),
                L.Tr("statusLine 命令: ", "statusLine command: ") + (Command ?? L.Tr("（无）", "(none)")),
                L.Tr("透传的原命令: ", "Wrapped original command: ") + (WrappedInner ?? L.Tr("（无）", "(none)")),
                L.Tr("ratelimits.json: ", "ratelimits.json: ") + (RatelimitsExists ? L.Tr("存在", "present") : L.Tr("缺失", "missing")),
                CapturedAt is DateTime c
                    ? L.Tr("上次额度数据: ", "Last usage data: ") + c.ToString("yyyy-MM-dd HH:mm:ss")
                    : L.Tr("上次额度数据: 尚无", "Last usage data: none yet"),
                L.Tr("settings.json: ", "settings.json: ") + SettingsPath,
                L.Tr("支持目录: ", "Support directory: ") + SupportDirPath,
            };
            return string.Join("\n", lines);
        }
    }

    public static Diagnostics GetDiagnostics()
    {
        DateTime? captured = null;
        bool exists = File.Exists(Paths.RatelimitsFile);
        if (exists)
        {
            try
            {
                using var doc = JsonDocument.Parse(File.ReadAllText(Paths.RatelimitsFile));
                if (doc.RootElement.TryGetProperty("capturedAt", out var t) && t.TryGetDouble(out var secs))
                    captured = DateTimeOffset.FromUnixTimeSeconds((long)secs).LocalDateTime;
            }
            catch { }
        }
        return new Diagnostics(IsInstalled, CurrentCommand(), InnerCommand(), exists, captured,
            Paths.ClaudeSettings, Paths.SupportDir);
    }
}
